cedar-lean/Cedar/Validation/Typechecker.lean

```diff
@@ -157,16 +166,36 @@ def typeOfInₛ (ety₁ ety₂ : EntityType) (x₁ x₂ : Expr) (env : Environme
     then .anyBool
     else .ff
 
+def typeOfHasTag (ety : EntityType) (x : Expr) (t : Expr) (c : Capabilities) (env : Environment) : ResultType :=
+  match env.ets.tags? ety with
+  | .some .none     => ok (.bool .ff)
+  | .some (.some _) =>
+    if (x, .tag t) ∈ c
+    then ok (.bool .tt)
+    else ok (.bool .anyBool) (Capabilities.singleton x (.tag t))
+  | .none           =>
+    if actionType? ety env.acts
+    then ok (.bool .ff) -- action tags not allowed
+    else err (.unknownEntity ety)
+
+def typeOfGetTag (ety : EntityType) (x : Expr) (t : Expr) (c : Capabilities) (env : Environment) : ResultType :=
+  match env.ets.tags? ety with
+  | .some .none      => err (.tagNotFound ety t)
+  | .some (.some ty) => if (x, .tag t) ∈ c then ok ty else err (.tagNotFound ety t)
+  | .none            => err (.unknownEntity ety)
+
 def ifLubThenBool (ty₁ ty₂ : CedarType) : ResultType :=
   match ty₁ ⊔ ty₂ with
   | some _ => ok (.bool .anyBool)
   | none   => err (.lubErr ty₁ ty₂)
 
-def typeOfBinaryApp (op₂ : BinaryOp) (ty₁ ty₂ : CedarType) (x₁ x₂ : Expr) (env : Environment) : ResultType :=
+def typeOfBinaryApp (op₂ : BinaryOp) (ty₁ ty₂ : CedarType) (x₁ x₂ : Expr) (c : Capabilities) (env : Environment) : ResultType :=
   match op₂, ty₁, ty₂ with
   | .eq, _, _                               => typeOfEq ty₁ ty₂ x₁ x₂
   | .mem, .entity ety₁, .entity ety₂        => ok (.bool (typeOfInₑ ety₁ ety₂ x₁ x₂ env))
   | .mem, .entity ety₁, .set (.entity ety₂) => ok (.bool (typeOfInₛ ety₁ ety₂ x₁ x₂ env))
+  | .hasTag, .entity ety₁, .string          => typeOfHasTag ety₁ x₁ x₂ c env
```

Comment:
Is there a reason we have `GetAttr`/`HasAttr` as separate AST notes but not for tags?

Reply:
Ah it's that the rhs is an expression not an `Attr`, that makes sense

Reply:
Yes, that was the rationale.  Curious: did we choose to make them separate nodes in Rust?

Reply:
In Rust I added them to `BinaryOp` as well